### DISCUSSION TITLE
Custom recipes in separate composer package v6 vs v7 (autoloading issue)

### DISCUSSION THREAD

**Discussion by adrenth** [deployphp/deployer#3294] [General]
- Deployer version: 7.0.1
- Deployment OS: Linux

At v6 we had some packages containing custom recipes (basically a copy of like https://github.com/deployphp/recipes).
Unfortunately after upgrading to v7 this setup does not work anymore. 

The autoloading has been broken and we get messages like:

```text
 ErrorException in deploy.php on line 8:
  require_once(recipe/my-custom-recipe.php): Failed to open stream: No such file or directory
```

The `composer.json` file of the package contains this autoload configuration:

```json
    "autoload": {
        "files": [
            "autoload.php"
        ]
    }
```

The `autoload.php` file contains:

```php
set_include_path(get_include_path() . PATH_SEPARATOR . __DIR__);
```

The file structure of the package:

```
recipe/my-custom-recipe.php
composer.json
autoload.php
```

What is the best practice for adding custom recipes to projects (using a composer package)? Couldn't find any documentation on this subject.

**Comment by antonmedv** [comment]
What about the direct approach?

```php
require_once "vendor/my/recipes/recipe/my-custom-recipe.php";
```

**Comment by adrenth** [comment]
That could be a good alternative. However, in our situation we have included Deployer and the custom recipes in a Docker container which is responsible for deploying projects. Deployer and its dependencies are installed in the container as **global**. In v6 this would be fine, since the dependencies of composer global were included as well.

This was our strategy:

- Build or start a docker container
- Install composer
- composer require global [deployer/recipes/other]
- git clone [project repository]
- run "dep deploy ..."

We have not included Deployer and it's recipes and requirements to the project itself (because of maintenance). So that if Deployer or any of its requirements needed a hotfix, or a set of new recipes, we don't need to address this in hundreds of projects, but we only have to update the "Docker Deployment Container".

**Comment by loevgaard** [comment]
For anybody finding this discussion, here is a new issue describing this problem: https://github.com/deployphp/deployer/issues/3485
